### ISSUE TITLE
Load panel is broken on ORBIS

### ISSUE THREAD

**Issue by Backporter** [schlangster/skyui#91]
Skyui 2.2 works, anything above that does not work, videos showing the issue below:

[Normal load panel](https://cdn.discordapp.com/attachments/535530099475480596/957322558125006848/2022-03-26_12-57-25.mp4)
[Slyui's Load pannel](https://cdn.discordapp.com/attachments/535530099475480596/957322904415141919/2022-03-26_12-59-09.mp4)

**Comment by Mardoxx** [comment]
Interesting. I got it to work on PS3 a few years ago but decided not put
any particular effort in to it due to lack of features requiring SKSE.

I was never in a position to test this on PS4, impressed you got it working!

Seems odd how it doesn’t work after version 2.2. I can’t remember what
menus were modified in which versions. Judging from your videos it looks
like in 2.2 the Journal Menu (pause menu) wasn’t  modified.

As a quick fix you could extract the quest_journal.gfx from the PS4’s
update.bsa, and put that in an appropriate directory so it overrides
SkyUI’s.

As for the missing translations, if you extract all the translations from
SkyUI’s translation.txt files and copy paste this to Skyrim’s ones it
should pick them up.

The mechanism to include translations SkyUI uses in later versions requires
SKSE

I believe controller support and button art should work correctly, I do
remember ensuring it worked on PC with a 360 controller and ensured the
correct code was functional as much as I could should someone ever get it
working on a different platform.

**Comment by Mardoxx** [comment]
You may need to download the button art mod to make it work completely for
PlayStation buttons.

**Comment by Backporter** [comment]
I did actually tried keeping the original quest_journal.gfx from the game and it still causes the same error sadly, I also tried manually settings the load index to 1, and some other small things and the issue persisted, I ran out of idea's as this part of the game isn't within my skillset.

**Comment by Mardoxx** [comment]
How odd. Sounds like it’s not overriding it properly. With the original
journal menu it should not have the MOD CONFIGURATION menu item

**Comment by Mardoxx** [comment]
You could unpack the skyui bsa and delete the quest journal from it.

**Comment by Mardoxx** [comment]
Wait I think I remember. The override order is (highest to lowest)

Loose file swf
bsa/bsm load order swf
Loose file gfx
bsa/bsm load order gfx

So the loose/archive gfx is not getting overridden

I think you can change the file ext to swf and change the first three bytes
to WSF (instead of CWS or whatever it is for a gfx)

**Comment by Backporter** [comment]
repacking the Skyui BSA without the quest_journal does fix it, however ill still need the MCM

**Comment by Mardoxx** [comment]
I’m not sure sure the MCM will work! I think it requires hooks added in by
SKSE. I can’t remember for sure though, the approach taken was to use as
much of the functionality available from the base game as possible.

**Comment by Backporter** [comment]
I have my own library that adds support for the MCM and many other things, that's one of the main reasons I was hoping I could find a fix for this that way its not a trade off, MCM vs loading/saving

any example of my library working with one of the [coolest mods ever](https://www.youtube.com/watch?v=Gcakb1Gdwb0)

**Comment by Mardoxx** [comment]
I don’t have any of my old tools for decompiling the UI any more. Would be
interesting to see if the PS4 journal used a different mechanism from the
PC UI to save/load and that is perhaps why it isn’t working.

I wouldn’t be able to say for sure, but you can decompile the UI files with
any flash decompiled

**Comment by Backporter** [comment]
I've been using JPEXS to view the PS4 swf's, awhile ago I had noticed that skyui's swf was missing a function call along with some other stuff and had assumed that's what the issue was, but my action script and skyrim's UI stuff is not something i normally work on so it was just a guess, I can post the PS4's swf's if that help?

**Comment by Backporter** [comment]
overwrote the game's default logger from the GFxLoader part of the game with my own so I can see what's going on, then opened up the save/load panel and tried to load so it'd print out more info, closed the game and opened up the log, this is all I found regarding Save/Load inside the log

``Error: CallMethod - '_level0.QuestJournalFader.Menu_mc.SystemFader.Page_mc.SaveLoadPanel.List_mc.selectedIndex' is not a method``

**Comment by Backporter** [comment]
https://youtu.be/GC2VyF1T0cg what lol, what all code logic does the MCM require?

**Comment by Mardoxx** [comment]
I suspect it requires some additional functionality added by SKSE.

**Comment by Backporter** [comment]
yeah, don't suppose there's a breakdown on how skyui's MCM actually works?, I've implemented the min that still allows the MCM to function on PC, yet no dice, it says the menus worked, but they are missing from the list.

**Comment by Mardoxx** [comment]
If you look at SKI_ConfigManager line 116

When the journal menu is opened it calls setModNames with the mod names
(sent as varargs array) on the config panels movie clip

ConfigPanel.as line 170

**Comment by Mardoxx** [comment]
Then when you select a menu item it does similar but calling setPageNames
and then when you select a page name it does some magic with writing
options to some array buffer of options and option types

**Comment by Mardoxx** [comment]
So you’ll need to implement SKSE’s papyrus UI interop module at the very
least I think

**Comment by Backporter** [comment]
I already implemented entire PapyrusUI stuff, same with "events"/"queues", I'm guessing it's needing names or something so ill go implement that I guess

**Comment by Backporter** [comment]
that was quick to implement, but nop, same thing :(

**Comment by Mardoxx** [comment]
Try logging some stuff in the OnMenuOpen see if that’s actually getting
called

**Comment by Backporter** [comment]
`Events.h 36 🤔 Menu "Journal Menu" IsOpening = "TRUE"`
`Events.h 36 🤔 Menu "Fader Menu" IsOpening = "FALSE"`

it's def being invoked

**Comment by Backporter** [comment]
also, not sure what the issue is on this regard, but activeeffects.swf actively triggers a crash lol

**Comment by Backporter** [comment]
same with any other mod that adds active UI elements into the huh, like AMatterOfTime for example, instant crash.

**Comment by Mardoxx** [comment]
Is papyrus listening for these events?
Is it successfully handling them?

**Comment by Mardoxx** [comment]
Whenever the game would crash with Ui components it was usually due to it
trying to load a file which didn’t exist (it was looking in the wrong place)

I vaguely remember some weird case where it would be trying to load like a
gfx from the exported(?) folder (the one with all the gfx files in it) if
you had the menu exported as a gfx?

**Comment by Mardoxx** [comment]
If it is this, it would be not impossible to patch it to look in the
correct place I imagine

**Comment by Backporter** [comment]
Is surely looks like papyrus is handling them.

the mod was left in exactly like it was when i downloaded it from nexus, here's some logs that are kinda useless as you'll see below
`[OrbisLog::Write] [WidgetLoader.as: loadWidget(a_widgetID = 2, a_widgetSource = AMOTClockWidget.swf)]`
`[OrbisLog::WriteVA] [Error, BSScaleformFileOpener failed to open Interface/Exported/widgets/AMOTClockWidget.gfx]`
`[OrbisLog::WriteVA] [GFxLoader failed to open "Interface/Exported/widgets/AMOTClockWidget.gfx"`

the thing about the swf loading is the PS4 version of the game is a liar and says it couldn't load a swf but it did in fact load(widgetloader.swf being a example of this, even when it fails to load i can see it loaded perfectly fine)

**Comment by Backporter** [comment]
![Skyrim_20220513093443](https://user-images.githubusercontent.com/29647778/168293884-8317342a-7d64-43e3-bf77-21aa0c591a96.jpg)

;)))

**Comment by Mardoxx** [comment]
Fantastic, what was the fix?

**Comment by Backporter** [comment]
one of the event queues was being ignored for a silly reason, I fixed that and now it's showing, mod pages are empty through lol

![Skyrim_20220513094231](https://user-images.githubusercontent.com/29647778/168295092-bfe71727-f45b-4755-ae14-7c92d3b01d2b.jpg)

**Comment by Mardoxx** [comment]
Well it’s progress! I’d think this is a similar issue it’s all done with
handling events sent from the papyrus scripts

**Comment by Backporter** [comment]
I'm wondering if AMOT would work now 🤔 will test.

**Comment by Backporter** [comment]
scratch that, instant crash.

**Comment by Backporter** [comment]
implemented more stuff for the MCM, now it crashes when you try and go into a mod panel, it crashes the same way when new UI elements are added interesting enough, I wonder if if maybe one of the file formats in the swf is PC specific?

**Comment by Backporter** [comment]
![Skyrim_20220516020959](https://user-images.githubusercontent.com/29647778/168529342-e6e70dfc-125f-4a21-9993-90a78f37fbc3.jpg)

https://youtu.be/qSYDeW5YTps

**Comment by Mardoxx** [comment]
Very nice! Do the options work and persist in the save?

**Comment by Backporter** [comment]
yes to both of those questions, using AMOT to confirm shows both work fine

**Comment by Mardoxx** [comment]
That’s amazing, great job!

How did you bake all the SKSE functionality into the elf? That’s really
impressive

**Comment by Backporter** [comment]
I actually put all this logic into a PRX using OpenOrbis SDK and use mira/ps4debug to load the PRX into the games process, I just had to reverse a lot of game stuff and make sure that matched, was surprisingly easy once I got past some weirdness that the PS4 version has lol

**Comment by Mardoxx** [comment]
Ahh very neat!

SkyUI should work correctly with the controller - this is something I made
sure of for the PC version. You can override the button art for a
PlayStation controller by replacing the buttonArt swf with the PlayStation
one (it’s on nexus somewhere)

**Comment by Mardoxx** [comment]
You can load this up as a separate mod as long as it comes after SkyUI - so
you don’t need to repackage SkyUI

**Comment by Backporter** [comment]
you mean [this](https://www.nexusmods.com/skyrim/mods/29450)? if so I have it installed, it only seems to effect specific keys(mainly controller specific keys) and not all of them, I'm guessing the fact that the PS4 uses a totally different set of keycodes is why this occurs

**Comment by Mardoxx** [comment]
Yes it should just replace any Xbox keys with the ps4 ones. Other ones are
the same.
Have they mapped like the key code to Enter to X or something?

**Comment by Backporter** [comment]
the PS4 keycodes are sony keycodes iirc, so they would be [these keycodes](https://github.com/OpenOrbis/OpenOrbis-PS4-Toolchain/blob/master/include/orbis/_types/pad.h#L7) rather than directx ones

**Comment by Mardoxx** [comment]
Ah I’m not sure then. Also the touchpad is new isn’t it so that probably
won’t work in skyui anyway

**Comment by Backporter** [comment]
I can likely translate the DirectX keycodes to Sony ones and fix it, that is assuming that is the issue, funny enough the touchpad works in the inventory system, if I hit it, it swaps to the magic panel and vice-versa

**Comment by Mardoxx** [comment]
Interesting, it must map to Tab some how!

**Comment by Backporter** [comment]
indeed, when I was reversing the Scalefrom/GFx stuff, I didn't find direct mouse support, so I'm assuming it was mostly removed

**Comment by Mardoxx** [comment]
Does the barter menu still work (switching between you and the vendor)?

I think L3 (or R3?) changes sort order and some other button changes the
sort column. You can also open the column filter menu with some other
button too. It’s written down somewhere I can’t remember.

**Comment by Backporter** [comment]
I swapped those keycodes in skyui's INI to the touchpad + R3/L3 keycodes as the default codes wouldn't work

**Comment by Backporter** [comment]
https://youtu.be/F5ho2le341E

**Comment by Mardoxx** [comment]
Excellent!

**Comment by Backporter** [comment]
darn, I really don't know what this issue is, whenever I go into a MCM that has options you can select it makes the MCM go wonky and show the translation rather than the key, has this been encountered before?

https://youtu.be/DWGDNSw3j2c

**Comment by Mardoxx** [comment]
No I haven't, is the same replicable on PC wit AMOT installed?

Does it happen with options selectors in other mods or just AMOT?

Perhaps it's a bug in AMOT?

**Comment by Backporter** [comment]
I Haven't seem it occur on PC, this can happen on other mods that has the same type of setting that is shown in that video

**Comment by Mardoxx** [comment]
Hmmm I think it’s a bug with their integration with the MCM, possibly in
the event handler for the on item selected (or whatever it is called)
they’re not running the text through the SKSE translation function?

**Comment by Backporter** [comment]
It looks fine from what I saw, it seems to just be that specific setting type that triggers this, I've never this happen on any other setting, I confirmed that the impl of the custom translations looked fine so unsure what is causing this.

**Comment by Mardoxx** [comment]
Does it do it on the pc version too

**Comment by Backporter** [comment]
no

**Comment by Mardoxx** [comment]
Strange! I’m not sure then

**Comment by Backporter** [comment]
I tried to do the same thing as SKSE for the deconstructor instead of what i normally do for GFXValue and it seems to have fixed the $ stuff, but looking at the log makes it not look like a good idea...

> ActionScript Memory leaks in movie 'Interface/MessageBox.swf', including 822 string nodes
> Leaked string content: 'skse', 'ExtendForm', 'ExtendData', 'ForceContainerCategorization', 'EnableMapMenuMouseWheel', 'ExtendAlchemyCategories', 'ShowOnMap', 'RequestActorValues', 'RequestActivePlayerEffects', 'StartRemapMode', 'LoadIndices', 'StoreIndices', 'SendModEvent', 'GetModList', 'AllowTextInput', 'GetPlayerSex'

seems the UI stuff on the PS4 is not a fan of adding stuff lol

**Comment by Backporter** [comment]
figured out what the issue was with both, now the only thing i need to fix is the abilty to bind gamepad buttons directly in the UI rather than needing to edit the INI/keycodes directly

**Comment by Mardoxx** [comment]
I would try logging in the OnMenuOpen and recompile SkyUI’s papyrus source
to make sure

I think what is going on with the swf crashing is what I tried to outline
above

This is because when the menu is an swf it tries to load relative paths
from Interface/path/to/blah.swf

But if it’s a gfx it tries to load from Interface/Exported/path/to/blah.swf

Hope that helps

**Comment by Backporter** [comment]
everything for the most part is working, no crashing, etc, the inability to load a save from the system page is the only thing that is actually broken and I don't think I can fix it myself

the whole "ActionScript Memory leaks in movie" was caused by a bad implementation of ~GFxValue and I resolved it
the MCM panels glitching and showing $XXXXX rather then the real text was also caused by ^
the magic/item menu crashing was also caused by ^ and I resolved it

the warnings from BSScaleformFileOpener are normal(they show up on the PC version too if you enable GFx Logging in SKSE, whenever I was testing to see if they showed up on PC, for some reason my SKSE install got messed up so I reinstalled it and they started to show up)

I've implemented the favorites menu set stuff so the filter works, ~~but this does actually lead to a crash if you try and select anything in it, currently trying to find a fix~~ (***FIXED***)

![Skyrim_20220813025718](https://user-images.githubusercontent.com/29647778/195163749-8f67da70-2085-47bc-85d0-2bc5089085cb.jpg)
in this screenshot you can see ActiveEffects now, you can also see AMOT fully working

I also designed(with lots of help from many many people in the skyrim modding scene)[ a Custom SWF showing the framerate for the PS4](https://www.youtube.com/watch?v=3ufSGpU3zgA&ab_channel=GamerSixVideos)

**Comment by Backporter** [comment]
is there any specific reason why location finder in the map menu is explicitly disabled if a controller is detected?

**Comment by Mardoxx** [comment]
There was a reason yes, I can’t remember though… either there wasn’t a
button free to get to the search box, or no means for inputting (which
wouldn’t involve another game code hack)

**Comment by Backporter** [comment]
I see, are the bumpers not mappable? those would be great buttons for these tbh, there is actually a native way to get keyboard inputs on PS4 at least, "ShowVirtualKeyboard" opens the typical PS4 keyboard you see everywhere and gives you the output of it, and using a USB keyboard(the console menu does this, never looked into how).

**Comment by Mardoxx** [comment]
Yeah this is what I wanted to do, but plugging in a controller to the PC
disables the keyboard I think so maybe that is why I disabled it. Because
then you have no means of input!

**Comment by Backporter** [comment]
![Skyrim_20221122060046.jpg](https://cdn.discordapp.com/attachments/957300345694986331/1044567348910227526/Skyrim_20221122060046.jpg)

may or may not lead to a crash for whatever reason, still missing text input support, but it does at least highlight the marker, better then nothing i suppose.

how: 
set MapMenu.as#454 - 457 to false,  
set MapMenu.as#459 & 460 to true, 
set MapMenu.as#377 to `if (true)`, 
set MapMenu.as#378 to `if (GlobalFunc.IsKeyPressed(details) && (details.skseKeycode == 4)) `

**Comment by Backporter** [comment]
I know this is a old thread and whatnot but I thought I'd post what I've done/found just in case someone finds it useful, so, I've got it somewhat fixed I suppose, instead of trying to figure out what exactly is wrong inside the SWF's code wise, I've went about patching the games `IsOKtoLoad` to always return OK and saving works(somewhat, the game only shows the saves for your active character, but it's better then nothing).
